tests/johnson_alahi_li_2016/test_modules.py

```diff
@@ -0,0 +1,351 @@
+from torch.hub import load_state_dict_from_url
+
+import pystiche
+import pystiche_papers.johnson_alahi_li_2016 as paper
+from pystiche.misc import to_2d_arg
+from pystiche_papers.johnson_alahi_li_2016 import modules
+from pystiche_papers.utils import ResidualBlock
+
+from .._asserts import assert_downloads_correctly, assert_is_downloadable
+
+
+@pytest.fixture(scope="module")
+def model_url_configs(styles):
+    return (
+        {
+            "framework": framework,
+            "style": style,
+            "impl_params": impl_params,
+            "instance_norm": instance_norm,
+        }
+        for framework, style, impl_params, instance_norm in itertools.product(
+            ("pystiche", "luatorch"), styles, (True, False), (True, False)
+        )
+    )
+
+
+def model_url_should_be_available(framework, style, impl_params, instance_norm):
+    # TODO: remove when pystiche weights are uploaded
+    if framework == "pystiche":
+        return False
+
+    if framework == "luatorch" and not impl_params:
+        return False
+
+    if (
+        framework == "luatorch"
+        and style in ("composition_vii", "starry_night", "the_wave")
+        and instance_norm
+    ):
+        return False
+
+    if (
+        framework == "luatorch"
+        and style in ("candy", "feathers", "mosaic", "the_scream", "udnie",)
+        and not instance_norm
+    ):
+        return False
+
+    return True
+
+
+@pytest.fixture(scope="module")
+def model_urls(model_url_configs):
+    return tuple(
+        modules.select_url(**config)
+        for config in model_url_configs
+        if model_url_should_be_available(**config)
+    )
+
+
+def test_select_url(subtests, model_url_configs):
+    for config in model_url_configs:
+        with subtests.test(**config):
+            if model_url_should_be_available(**config):
+                assert isinstance(modules.select_url(**config), str)
+            else:
+                with pytest.raises(RuntimeError):
+                    modules.select_url(**config)
+
+
+@pytest.mark.slow
+def test_weights_downloadable(subtests, model_urls):
+    for url in model_urls:
+        with subtests.test(url):
+            assert_is_downloadable(url)
+
+
+@pytest.mark.large_download
+@pytest.mark.slow
+def test_weights_download_correctly(subtests, model_urls):
+    for url in model_urls:
+        with subtests.test(url):
+            assert_downloads_correctly(url)
+
+
+def test_get_conv(subtests):
+    in_channels = out_channels = 3
+    kernel_size = 3
+    stride = 1
+    for padding, upsample in itertools.product((None, 1, (1, 1)), (True, False)):
+        conv_module = modules.get_conv(
+            in_channels,
+            out_channels,
+            kernel_size,
+            stride=stride,
+            padding=padding,
+            upsample=upsample,
+        )
+
+        assert isinstance(
+            type(conv_module),
+            type(nn.ConvTranspose2d) if upsample else type(nn.Conv2d),
+        )
```

Comment:
Please review [the difference between `type()` and `isinstance()`](https://www.guru99.com/type-isinstance-python.html#6). Afterwards, apply this change to every occurrence:

```suggestion
        assert isinstance(
            conv_module,
           nn.ConvTranspose2d if upsample else nn.Conv2d,
        )
```